mlflow/pyfunc/__init__.py

```diff
@@ -3045,57 +3043,62 @@ def predict(self, context, model_input: List[str], params=None) -> List[str]:
-                        input_example=input_example,
+        # only infer signature based on input example when signature
+        # and type hints are not provided
+        if signature is None and signature_from_type_hints is None:
+            saved_example = _save_example(mlflow_model, input_example, path, example_no_conversion)
+            if saved_example is not None:
+                _logger.info("Inferring model signature from input example")
+                try:
+                    mlflow_model.signature = _infer_signature_from_input_example(
+                        saved_example,
+                        _PythonModelPyfuncWrapper(model_for_signature_inference, None, None),
+                    )
+                except Exception as e:
+                    _logger.warning(
+                        f"Failed to infer model signature from input example, error: {e}",
                     )
-            # only infer signature based on input example when signature
-            # and type hints are not provided
-            if signature is None and signature_from_type_hints is None:
-                if saved_example is not None:
-                    _logger.info("Inferring model signature from input example")
-                    try:
-                        context = PythonModelContext(artifacts, model_config)
-                        python_model.load_context(context)
-                        mlflow_model.signature = _infer_signature_from_input_example(
-                            saved_example,
-                            _PythonModelPyfuncWrapper(python_model, None, None),
-                        )
-                    except Exception as e:
-                        _logger.warning(
-                            f"Failed to infer model signature from input example, error: {e}",
-                        )
-                    else:
-                        if type_hint_from_example:
-                            update_signature_for_type_hint_from_example(
-                                input_example, mlflow_model.signature
-                            )
                 else:
-                    if type_hint_from_example:
-                        _logger.warning(
-                            _TYPE_FROM_EXAMPLE_ERROR_MESSAGE,
-                            extra={"color": "red"},
-                        )
-                    # if signature is inferred from type hints, warnings are emitted
-                    # in _infer_signature_from_type_hints
-                    elif not should_infer_signature_from_type_hints:
-                        _logger.warning(
-                            "Failed to infer model signature: "
-                            f"Type hint {type_hints} cannot be used to infer model signature and "
-                            "input example is not provided, model signature cannot be inferred."
+                    if type_hint_from_example and mlflow_model.signature:
```

Comment:
With the existing code, an unexpected exception is raised when `TypeFromExample` is specified and signature cannot be inferred from input example (e.g. ["a", 0])